ivy: complex number support

glxxyz opened the pull request:
An idea for complex number support in ivy. Uses a Go-like complex number syntax instead of APL's `J`. Go's complex numbers weren't used for the implementation, instead the real and imaginary parts can be any ivy number type.

Adds 3 unary operators real, imag, and phase. An `i` suffix that allows complex numbers to be input e.g. `2+3i`. 
 The `value.Complex` implementation could be improved by removing some of the EvalUnary/EvalBinary; this was quick to implement and it maintains the number types wherever possible. Tests aren't exhaustive- the whole approach may be a non-starter.

This may not be something that's even desirable to have in ivy, I noticed a mention of 'imaginary' in scan.go though so maybe complex number support was considered at one time? https://github.com/robpike/ivy/blob/c38c34e9eb24ab33e88f1a92ff5d825b32296e6b/scan/scan.go#L451

robpike commented:
This is an interesting idea, but I am not sure it's a worthwhile path to take. To do it, it must be done well, which is a lot of work. The transcendental functions and logarithms must work, not to mention the surprising challenge of division.

But leaving aside whether to do it at all, this is not the best way to introduce them. The parsing problems introduced by using Go's notation here are severe.

On my morning constitutional I thought about the problem and realized that the right APL way (and hence ivy way) is to use operators. I thought of 'i' for the operator, but that is stealing a very common variable, so I decided that 'j' would be better, as the electrical engineers use.  Then the unary 'j' returns an imaginary number, or perhaps just multiplies its argument by sqrt(-1). The binary 'j' takes a  real and an imaginary part:

```
j 3 == (0+3i)
2 j 4 == (2+3i)
```

This has a strong analogy with how rationals are built, by overloading (but not really) the / operator. As with rationals, the lexer could promote the form directly and avoid most of the parsing mess.  Then you get things like this to work well:

```
j iota 3 == (0+1i) (0+2i) (0+3i)
2 j iota 3 = (2+1i) (2+2i) (2+3i)
1j1 2j2 3j3 == (1+1i) (2+2i) (3+3i)
```

and so on. If you executed those, you wouldn't get Go notation of course:

```
j iota 3 
   0j1 0j2 0j3

2 j iota 3
   2j1 2j2 2j3

1j1 2j2 3j3
   1j1 2j2 3j3
```


When I got home, I looked up what dyalog did, and at least for syntax, guess what? 2J3 is their notation for 2+3i. So that is an endorsement of my suggestion.

But to go back to the main point, thank you for suggesting the idea, but this is not the way to go, and I'm not sure it's worth doing at all, because the work involved is substantial. Introducing a new basic type into ivy is hard enough; doing it for such an unusual type is all the more work.

So I am going to close this PR, but I promise to think hard about doing it.

glxxyz commented:
Thanks for taking a look, I had already implemented complex division. Yes some of the transcendentals would be a significant amount of work. It wasn't difficult to add the basic interaction with other number types because they can all be promoted to Complex.

I switched to the 'j' operator and separator on my fork and your examples work. I'll keep refining it on my own fork. Here are few of the tests:

```
j 3
	0j3

2 j 4
	2j4

2/4j3/9
	1/2j1/3

j iota 3
	0j1 0j2 0j3

2 j iota 3
	2j1 2j2 2j3

1j1 2j2 3j3
	1j1 2j2 3j3

sqrt -2
	0j1.41421356237

2j3 / 5j7
	31/74j1/74

log 1j1
	0.34657359028j0.785398163397

1j1 log 2j2
	1.32596797216j-0.738702122273

phase 1j1
	0.785398163397

abs 1j1
	1.41421356237

floor 1.1j2.2
	1j2
```

robpike commented:
You triggered me making this change myself. I decided it would be fun. I didn't borrow or even look at your code, as I really wanted the joy of doing it myself. Hope that's OK.
I still need to do log and non-integer power, as well as the arctrigs and hyperbolics, but the foundations are there.
I also fixed a ton of bugs this exposed that were already there.

Thanks for suggesting it.
